File: zeddihub_v2/src/lib/api/http.ts
/**
 * Typed wrappers around Tauri http_* commands.
 *
 * Backend: src-tauri/src/commands/http.rs (TTL cache).
 */

import { invoke } from "@tauri-apps/api/core";

export const httpApi = {
  /**
   * Fetch JSON from URL with TTL cache. If `forceRefresh` is true,
   * bypasses cache. Returns cached payload on network failure if available.
   */
  fetchJson: <T = unknown>(url: string, ttlSeconds = 1800, forceRefresh = false) =>
    invoke<T>("http_fetch_json", { url, ttlSeconds, forceRefresh }),

  /** Returns last-fetched age in seconds, or null if never fetched. */
  cacheAge: (url: string) => invoke<number | null>("http_cache_age", { url }),

  /** Drop a single URL from cache (force re-fetch next time). */
  invalidate: (url: string) => invoke<void>("http_invalidate", { url }),
};

File: src-tauri/src/commands/http.rs
//! Tauri commands — thin wrappers around `services::http_cache`.

use tauri::State;
use crate::error::AppResult;
use crate::services::http_cache::HttpCache;

#[tauri::command]
pub async fn http_fetch_json(
    state: State<'_, HttpCache>,
    url: String,
    ttl_seconds: u64,
    force_refresh: bool,
) -> AppResult<serde_json::Value> {
    state.fetch_json(&url, ttl_seconds, force_refresh).await
}

#[tauri::command]
pub async fn http_cache_age(
    state: State<'_, HttpCache>,
    url: String,
) -> AppResult<Option<u64>> {
    Ok(state.age_seconds(&url).await)
}

#[tauri::command]
pub async fn http_invalidate(
    state: State<'_, HttpCache>,
    url: String,
) -> AppResult<()> {
    state.invalidate(&url).await;
    Ok(())
}
